### PR TITLE
use AZURE_ASSERT in storage code

### PR DESCRIPTION
**Pull request by Jinming-Hu**
# Pull Request Checklist

Please leverage this checklist as a reminder to address commonly occurring feedback when submitting a pull request to make sure your PR can be reviewed quickly:

See the detailed list in the [contributing guide](https://github.com/Azure/azure-sdk-for-cpp/blob/master/CONTRIBUTING.md#pull-requests).

- [x] [C++ Guidelines](https://azure.github.io/azure-sdk/cpp_introduction.html)
- [x] Doxygen docs
- [x] Unit tests
- [x] No unwanted commits/changes
- [x] Descriptive title/description
  - [x] PR is single purpose
  - [x] Related issue listed
- [x] Comments in source
- [x] No typos
- [x] Update changelog
- [x] Not work-in-progress
- [x] External references or docs updated
- [x] Self review of PR done
- [x] Any breaking changes?

### CODE REVIEW

File: sdk/storage/azure-storage-blobs/src/blob_lease_client.cpp [hunk 1]
```diff
@@ -52,12 +53,16 @@ namespace Azure { namespace Storage { namespace Blobs {
...
-      }
+      AZURE_ASSERT_MSG(
+          !options.AccessConditions.IfMatch.HasValue(),
+          "Blob container lease doesn't support If-Match condition");
+      AZURE_ASSERT_MSG(
+          !options.AccessConditions.IfNoneMatch.HasValue(),
+          "Blob container lease doesn't support If-None-Match condition");
+      AZURE_ASSERT_MSG(
+          !options.AccessConditions.TagConditions.HasValue(),
+          "Blob container lease doesn't support tag condition");
```
Comment (vhvb1989) [inline sdk/storage/azure-storage-blobs/src/blob_lease_client.cpp:64]: Don't use ASSERT_MSG, use ASSERT alone.  ASSERT_MSG is only for when the condition is well known to be false, like the non_implemented macro

**Review by vhvb1989 — COMMENTED** [review]
Use AZURE_ASSERT() instead of AZURE_ASSERT_MSG

File: sdk/storage/azure-storage-blobs/src/blob_lease_client.cpp (hunk 1 above)
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-blobs/src/blob_lease_client.cpp:64]: But `AZURE_ASSERT` doesn't print a message, which can be helpful to customers if they hit an assertion failure.

**Comment by Jinming-Hu** [comment]
> Use AZURE_ASSERT() instead of AZURE_ASSERT_MSG

What's the problem with `AZURE_ASSERT_MSG`. I think it's good

File: sdk/storage/azure-storage-blobs/src/blob_lease_client.cpp (hunk 1 above)
Comment (vhvb1989) [reply sdk/storage/azure-storage-blobs/src/blob_lease_client.cpp:64]: ASSUTE_ASSERT will print a message which will include the condition, file, line number, etc.  So, people will see this in that message:  `!options.AccessConditions.TagConditions.HasValue()` which is good enough to understand that, as a contract, tagConditions can't be there when calling that method.  The message would be mostly ignored or redundant.

The message is relevant when the assert is just something like `false`,  in that case, the message we get would contain only the word `false`.  That's the case where the message would help and useful.

Also, we want to call AZURE_ASSERT_MSG with `false` directly so we are sure that the message will be used.  If we pass a condition, and the condition is true, the message is just ignored and never used, but we are still building it as part of the Debug build.  

On release build, no matter if you pass a message or not, the ASSERT is ignored to just call abort() directly. So we won't see any message or anything on Production.

**Comment by vhvb1989** [comment]
> > Use AZURE_ASSERT() instead of AZURE_ASSERT_MSG
> 
> What's the problem with `AZURE_ASSERT_MSG`. I think it's good.
> 
> If it's not for internal or public use, you could've defined it in .cpp file.

I am not sure if I can move it to cpp.  It is the implementation of UNREACHABLE_CODE,

**Review by vhvb1989 — APPROVED** [review]
Fine if you want to keep the message, but you might want to try and see the console output with and without the message first. You might notice it could be a little redundant and not passing the string would be fine.

**Pull request merged**